opcm: Update OPCM.deploy to use v2 games

stevennevins opened the pull request:
**Description**

Update `OPCM.deploy` to deploy the default permissioned dispute game with the `PermissionedDisputeGameV2` implementation instead of blueprints when the v2 dispute game feature flag is on. 

**Metadata**

Closes #17257

stevennevins committed:
feat: integrate v2 implementations into opcm

stevennevins committed:
fix: skip v2 implementations if not deployed

stevennevins committed:
chore: bump semver version

stevennevins committed:
feat: conditionally set the v2 games based on feature flag

stevennevins committed:
feat: add override for setup to re-deploy opcm with feature flag

stevennevins committed:
test: improve test by deploying with the feature flag and asserting on output

stevennevins committed:
test: add assertion that the deployment configuration on the dgf is correct

stevennevins committed:
chore: improve test name

stevennevins committed:
refactor: add helper function for feature toggled deployment of opcm

stevennevins committed:
fix: bytecode size check failing

stevennevins committed:
chore: run forge fmt

stevennevins committed:
fix: address CI naming and semver bump

stevennevins committed:
fix: rename setup contract name to align with ci checks

mbaxter commented on packages/contracts-bedrock/interfaces/L1/IOPContractsManager.sol:
```diff
@@ -216,6 +221,8 @@ interface IOPContractsManager {
         address anchorStateRegistryImpl;
         address delayedWETHImpl;
         address mipsImpl;
+        address faultDisputeGameV2Impl;
+        address permissionedDisputeGameV2Impl;
```
We'll eventually drop the V2 - so I think it's less clean-up work to avoid it where we can:
```suggestion
        address faultDisputeGameImpl;
        address permissionedDisputeGameImpl;
```

stevennevins replied:
i think based on our discussion in the prior pr will leave this as something we come back to

stevennevins committed:
chore: comment out added fix

stevennevins committed:
fix: verify opcm

stevennevins committed:
fix: naming convention for test contracts

mbaxter commented on packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol:
```diff
@@ -398,6 +400,14 @@ contract VerifyOPCM is Script {
+        if (_isV2Implementation(_target.name)) {
+            console.log("  Status: [SKIP] V2 implementation, not yet deployed");
+            console.log(string.concat("  Status: [SKIP] Skipping verification for ", _target.name));
+            return true;
```
Can we use the `devFeatureBitmap` field on the OPCM to figure out whether we should validate these fields?  And should we be running these tests with different dev features enabled?

stevennevins replied:
yeah good suggestion

stevennevins replied:
Updated [5fda712](https://github.com/ethereum-optimism/optimism/pull/17408/commits/5fda712204c8c743098ca690451eecb49ae2c4e4)

stevennevins committed:
fix: move the updates for addGameType to the addGameType pr

mbaxter commented on packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol:
```diff
@@ -811,6 +821,17 @@ contract VerifyOPCM is Script {
+    /// @notice Checks if a contract name corresponds to a V2 implementation.
+    /// @param _contractName The contract name to check.
+    /// @return True if this is a V2 implementation, false otherwise.
+    function _isV2Implementation(string memory _contractName) internal pure returns (bool) {
```
(nit) It's not very clear what "v2" means here, maybe something like:
```suggestion
    function _isV2DisputeGameImplementation(string memory _contractName) internal pure returns (bool) {
```

mbaxter commented on packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol:
```diff
@@ -877,6 +898,13 @@ contract VerifyOPCM is Script {
+        string memory disputePath =
+            string.concat("forge-artifacts/", sourceName, ".sol/", _contractName, ".dispute.json");
+        if (vm.exists(disputePath)) {
+            return disputePath;
```
What's the effect of returning these `.dispute.json` artifacts here?

stevennevins replied:
It's the name of the compiler restriction artifacts. I'm going to try removing this. i'm not sure if i missed where i was supposed to put an override or something but i was having an issue with the artifacts matching the expectation here

stevennevins replied:
Removed this check, seems like it was an intermittent thing after I did a just clean. It seemed like it was actually another symptom related to the ProtocolVersions compiler artifact we noticed the other day. will double check CI in a bit to make sure it's resolved

mbaxter commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -1055,6 +1057,13 @@ contract OPContractsManagerDeployer is OPContractsManagerBase {
             )
         );
 
+        // Store v2 implementation addresses if the feature flag is enabled
+        if (isDevFeatureEnabled(DevFeatures.DEPLOY_V2_DISPUTE_GAMES)) {
+            // Store v2 implementation addresses for later registration with DisputeGameFactory
+            output.permissionedDisputeGameV2 = IPermissionedDisputeGameV2(implementation.permissionedDisputeGameV2Impl);
+            output.faultDisputeGameV2 = IFaultDisputeGameV2(implementation.faultDisputeGameV2Impl);
+        }
```
Why do we need to set this on the `output`?  I think `output` is meant to represent newly deployed contracts.

stevennevins replied:
ahh yeah nice catch

stevennevins replied:
fixed 
[4e1b47b](https://github.com/ethereum-optimism/optimism/pull/17408/commits/4e1b47b8cc2089438dd3905a20ae730600e8bc1d)

mbaxter commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -1139,11 +1148,25 @@ contract OPContractsManagerDeployer is OPContractsManagerBase {
+            setDGFImplementation(
+                output.disputeGameFactoryProxy,
+                GameTypes.PERMISSIONED_CANNON,
+                IDisputeGame(address(output.permissionedDisputeGameV2))
```
Shouldn't the game implementations come from the`Implementations` struct?:
```suggestion
                IDisputeGame(getImplementations().permissionedDisputeGameV2)
```

stevennevins replied:
yup! fixed
[4e1b47b](https://github.com/ethereum-optimism/optimism/pull/17408/commits/4e1b47b8cc2089438dd3905a20ae730600e8bc1d)

mbaxter commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -1139,11 +1148,25 @@ contract OPContractsManagerDeployer is OPContractsManagerBase {
+                && address(output.permissionedDisputeGameV2) != address(0)
+        ) {
+            // Register v2 implementation for PERMISSIONED_CANNON game type
+            setDGFImplementation(
```
Don't we have to set the `gameArgs` here?

stevennevins replied:
Fixed!
[4e1b47b](https://github.com/ethereum-optimism/optimism/pull/17408/commits/4e1b47b8cc2089438dd3905a20ae730600e8bc1d)

mbaxter commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -1661,20 +1707,67 @@ contract OPContractsManager_Deploy_Test is DeployOPChain_TestBase {
+        assertEq(address(output.permissionedDisputeGameV2), address(0));
+        assertEq(address(output.faultDisputeGameV2), address(0));
+
+        // Check that v1 implementation is registered for PERMISSIONED_CANNON
+        address registeredImpl = address(output.disputeGameFactoryProxy.gameImpls(GameTypes.PERMISSIONED_CANNON));
+        assertEq(registeredImpl, address(output.permissionedDisputeGame));
```
I think we should try creating each game type so we can check that all of the immutable fields are correct for both `WithV2Flag` and `WithoutV2Flag` tests.

stevennevins replied:
sgtm

stevennevins replied:
Added here for the PermissionedGames [46502c9](https://github.com/ethereum-optimism/optimism/pull/17408/commits/46502c95e2e94dcc1c2f6ce59553753c0f1c0786)

mbaxter commented on packages/contracts-bedrock/test/scripts/VerifyOPCM.t.sol:
```diff
@@ -126,6 +137,11 @@ contract VerifyOPCM_Run_Test is VerifyOPCM_TestInit {
             VerifyOPCM.OpcmContractRef memory ref = refs[randomImplIndex];
 
+            // Skip V2 implementations (not yet deployed)
+            if (_isV2Implementation(ref.name)) {
```
(discussion) Can we check against the `DevFeatures` on `opcm` to determine whether this should be skipped?  And can we run these tests with the v2 flag on and off?

stevennevins replied:
i'm okay with either, was thinking it's not super important to validate now and just opted to skip. same reasoning for elsewhere where we are checking the `_isV2Implementation`

stevennevins replied:
Updated here to check if we should be skipping based on the bitmap
[5fda712](https://github.com/ethereum-optimism/optimism/pull/17408/commits/5fda712204c8c743098ca690451eecb49ae2c4e4)

stevennevins committed:
fix: proposer removed

stevennevins committed:
fix: rename function for clarity in VerifyOPCM script

Rename _isV2Implementation to _isV2DisputeGameImplementation for better clarity about what types of contracts this function checks for.

stevennevins committed:
fix: add proper gameArgs for V2 dispute game registration and fix setting output implementation on outupt struct

Inphi commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -362,8 +364,23 @@ abstract contract OPContractsManagerBase {
+        IDisputeGameFactory _dgf,
+        GameType _gameType,
+        IDisputeGame _newGame,
+        bytes memory _gameArgs
```
Let's overload `setDGFImplementation` to have one that takes `_gameArgs` as input and one without.

stevennevins replied:
I'm okay with your suggestion. I had received feedback from @mbaxter in a similar function in a test file, suggesting that it should be done this way rather than the way you're suggesting. I would be curious what her opinion is here

mbaxter replied:
I think overloading makes sense here, so we only have to update the setDGFImplementation calls we care about right now.  I also think we should be more explicit on which `setImplementation` call we use - maybe we could do something like:
```
function setDGFImplementation(
    IDisputeGameFactory _dgf,
    GameType _gameType,
    IDisputeGame _newGame,
    bytes memory _gameArgs
)
    internal
{
    if (v2Enabled) {
      _dgf.setImplementation(_gameType, _newGame, _gameArgs);
    } else {
      if (_gameArgs.length > 0) {
        revert SomeMeaningfulError();
      }
      _dgf.setImplementation(_gameType, _newGame);
    }
}
```

stevennevins replied:
Not sure we need this, at least for how we are using this function currently, it's inside conditional blocks where we are already doing the feature flag check

Inphi commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -553,7 +570,9 @@ contract OPContractsManagerGameTypeAdder is OPContractsManagerBase {
             // type already exists, then its implementation will be overwritten.
-            setDGFImplementation(dgf, gameConfig.disputeGameType, IDisputeGame(address(outputs[i].faultDisputeGame)));
+            setDGFImplementation(
+                dgf, gameConfig.disputeGameType, IDisputeGame(address(outputs[i].faultDisputeGame)), bytes("")
```
Shouldn't the AddGameType also support the creator pattern? Consider extending `AddGameInput` to provide this info.

stevennevins replied:
yeah that's in another PR. i updated the PR description to reflect this is just for updating the opcm.deploy function

Inphi commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -1727,6 +1769,9 @@ contract OPContractsManager is ISemver {
         IDelayedWETH delayedWETHPermissionedGameProxy;
         IDelayedWETH delayedWETHPermissionlessGameProxy;
+        // V2 dispute game contracts (deployed when DEPLOY_V2_DISPUTE_GAMES flag is set)
+        IFaultDisputeGameV2 faultDisputeGameV2;
+        IPermissionedDisputeGameV2 permissionedDisputeGameV2;
```
for my understanding: If the feature flag is set, then we want to have two deployments of the FDG using the blueprint and the new creator pattern. Is this correct?

stevennevins replied:
Yeah, thats what i thought was the simplest. we might come back and make it more restrictive so that contracts are selectively deployed based on the feature flag instead of only gating their usage based on the feature flag

Inphi replied:
Ya, that makes sense. I discussed doing this with mbaxter offline and wanted to confirm whether this is still the case.

Inphi reviewed (commented):
AddGameType should support DG v2. The OCPM upgrade function probably also needs updating. But I'm not certain what the upgrade target currently is.

OPCM.migrate doesn't support DG v2. And that's OK. But let's create a ticket to track this with a TODO referencing it in the migrate function.

stevennevins committed:
test: add helper functin for creating game proxies

stevennevins committed:
test: refactor helper functions for test and remove impl from output struct

stevennevins committed:
test: test args correct on created proxy games

stevennevins committed:
chore: remove unused helper for now

stevennevins committed:
fix: remove guard clause from local issue

stevennevins committed:
chore: bump semver

stevennevins committed:
chore: revert changes to ignore v2 implementations

stevennevins committed:
fix: check bitmap and contract name for ref

stevennevins committed:
test: with feature toggle on

stevennevins committed:
chore: forge fmt

stevennevins committed:
Revert "chore: forge fmt"

This reverts commit d029fdb8dd12445a839026972e3859be7aae1470.

stevennevins committed:
Revert "test: with feature toggle on"

This reverts commit a932eb608d25f9acf486ce49aeb5318742c666de.

stevennevins committed:
chore: forge fmt

stevennevins committed:
fix: test compiler restriction fix

stevennevins committed:
Revert "fix: test compiler restriction fix"

This reverts commit 50282cf4fbf32b2a3615a4d9537e0887436f682d.

stevennevins committed:
fix: compiler bump

stevennevins committed:
fix: typo

stevennevins committed:
test: add test for verify opcm with v2 dispute games deployed

stevennevins committed:
test: add skips for v1 tests if v2 deployed

stevennevins committed:
fix: skip standard validator until its implemented

stevennevins committed:
fix: skip addGameType until its implemented

stevennevins committed:
fix: skip updatePrestate until its implemented

stevennevins committed:
fix: remove diff in natspec

stevennevins committed:
chore: add TODO comments with issue tracking to skips

stevennevins committed:
fix: bump semver

stevennevins committed:
chore: bump semver

mbaxter committed:
Merge branch 'develop' into snevin/issue-17257/creator-pattern-opcm-deploy

 Conflicts:
	packages/contracts-bedrock/snapshots/semver-lock.json
	packages/contracts-bedrock/test/L1/OPContractsManager.t.sol

mbaxter committed:
Fix silent merge conflicts with develop

mbaxter committed:
Bump semver version

mbaxter committed:
Run semver-lock

mbaxter committed:
Reintroduce OPContractsManager_Version_Test

mbaxter committed:
Cleanup - revert comment change

mbaxter committed:
Remove v2 contracts from DeployOutput

mbaxter committed:
Remove unused imports

mbaxter committed:
Update DeployOPChain tests to run across feature flags (in progress)

mbaxter committed:
Update OPCM tests to work with FeatureFlags options

mbaxter committed:
Merge PermissionedDisputeGame tests, use feature flags

mbaxter committed:
Merge FaultDisputeGame tests, use feature flag in setup

mbaxter committed:
Remove redundant test

mbaxter committed:
Fix test name

mbaxter committed:
Fix anchorRootNotFound test - pass in extra data

mbaxter committed:
Extend deployment test to check more fields

mbaxter committed:
Add guards around new setImplementation call

mbaxter committed:
Add new error to OPCM interface

mbaxter committed:
Merge branch 'develop' into snevin/issue-17257/creator-pattern-opcm-deploy

mbaxter committed:
Run semver-lock

mbaxter committed:
Prefix var with underscore

mbaxter committed:
Remove unused imports

mbaxter committed:
Regenerate snapshots

mbaxter committed:
Pass opcm as an argument to _verifyOpcmContractRef

mbaxter committed:
Reorganize verifyOPCM v2 skip logic

mbaxter committed:
Merge branch 'develop' into snevin/issue-17257/creator-pattern-opcm-deploy

mbaxter commented on packages/contracts-bedrock/test/dispute/DisputeGameFactory.t.sol:
```diff
@@ -200,13 +202,26 @@ contract DisputeGameFactory_TestInit is CommonTest {
+            return setupFaultDisputeGameV2(_absolutePrestate);
+        } else {
+            return setupFaultDisputeGameV1(_absolutePrestate);
+        }
```
Since we're now running across various DevFeatures in CI, merge the FaultDisputeGameV2.t.sol into FaultDisputeGame.t.sol, similarly for PermissionedDisputeGameV2.t.sol and PermissionedDisputeGame.t.sol.

mbaxter commented on packages/contracts-bedrock/test/scripts/VerifyOPCM.t.sol:
```diff
@@ -120,12 +121,23 @@ contract VerifyOPCM_Run_Test is VerifyOPCM_TestInit {
+            // Skip V2 dispute games when feature disabled
+            bool isV2DisputeGame =
+                LibString.eq(ref.name, "FaultDisputeGameV2") || LibString.eq(ref.name, "PermissionedDisputeGameV2");
+            if (isV2DisputeGame && !v2FeatureEnabled) {
```
This test attempts to mutate deployed contract code and then run validations, since the contract isn't deployed when the feature is disabled we skip it here.  Similarly for the other skip in this file.

smartcontracts replied:
Same comment as above, we should probably just deploy the contracts no matter what, I don't see any downside in deploying the implementations. Cleans the code up and reduces branching logic.

mbaxter replied:
responded on the other comment [here](https://github.com/ethereum-optimism/optimism/pull/17408#discussion_r2421008077)

ajsutton commented on packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol:
```diff
@@ -176,9 +179,18 @@ contract VerifyOPCM is Script {
     /// @param _skipConstructorVerification Whether to skip constructor verification.
-    function runSingle(string memory _name, address _addr, bool _skipConstructorVerification) public {
+    function runSingle(
+        IOPContractsManager _opcm,
```
nit: Do we need to add This param to the function documentation above?

ajsutton commented on packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol:
```diff
@@ -385,12 +397,15 @@ contract VerifyOPCM is Script {
     /// @param _skipConstructorVerification Whether to skip constructor verification.
     /// @return True if the contract reference is verified, false otherwise.
     function _verifyOpcmContractRef(
+        IOPContractsManager _opcm,
```
nit: Add to doc?

mbaxter committed:
Add TODO to fix acceptance test

ajsutton commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -1914,20 +1959,73 @@ contract OPContractsManager_Deploy_Test is DeployOPChain_TestBase {
+        IOPContractsManager.Implementations memory impls = opcm.implementations();
+        bool expectEmptyV2Implementations = isV2 ? false : true;
+        assertEq(address(impls.permissionedDisputeGameV2Impl) == address(0), expectEmptyV2Implementations);
+        assertEq(address(impls.faultDisputeGameV2Impl) == address(0), expectEmptyV2Implementations);
```
Could we add a string reason to these asserts? Otherwise it can be very hard to know why the test failed unless you dig into the execution trace.

ajsutton reviewed (approved):
LGTM. Just a couple of nits.

smartcontracts commented on packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol:
```diff
@@ -176,9 +179,18 @@ contract VerifyOPCM is Script {
     /// @param _skipConstructorVerification Whether to skip constructor verification.
-    function runSingle(string memory _name, address _addr, bool _skipConstructorVerification) public {
+    function runSingle(
+        IOPContractsManager _opcm,
```
We should avoid adding this parameter to `runSingle`. `runSingle` is used to validate one-off contracts and doesn't really have anything to do with OPCM even though it uses the underlying functions. We can just pass `address(0)` into `_verifyOpcmContractRef` or something.

mbaxter replied:
removed

smartcontracts commented on packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol:
```diff
@@ -401,6 +416,20 @@ contract VerifyOPCM is Script {
         console.log(string.concat("  Expected Runtime Artifact: ", artifactPath));
 
+        // Check if this is a V2 dispute game that should be skipped
+        if (_isV2DisputeGameImplementation(_target.name)) {
```
I don't think there's any reason why we would need to avoid deploying these implementations if the feature is disabled. Simply deploying the implementations doesn't actually do anything. It would also mean we can get rid of all of this diff in VerifyOPCM

mbaxter replied:
I generally think its cleaner to fully isolate new features behind a feature flag.  It's easier to ensure we don't accidentally push changes to production prematurely, and gives us better assurances in tests.  It would also require a bit of extra clean up outside of this PR if we wanted to switch directions at this point.

smartcontracts commented on packages/contracts-bedrock/test/opcm/DeployOPChain.t.sol:
```diff
@@ -134,25 +138,26 @@ contract DeployOPChain_Test is DeployOPChain_TestBase {
+            Duration.unwrap(pdg.maxClockDuration()), Duration.unwrap(disputeMaxClockDuration), "PDG maxClockDuration"
         );
+
+        if (isDevFeatureEnabled(DevFeatures.DEPLOY_V2_DISPUTE_GAMES)) {
```
Should we be asserting that the game parameters have been set properly inside of the game args?

mbaxter replied:
We could, but I think its simpler / less brittle to create a game and check all of the fields are populated as expected, which [we do here](https://github.com/ethereum-optimism/optimism/blob/63409d3254a7d1a7803e7e842cd1969b03f432ad/packages/contracts-bedrock/test/L1/OPContractsManager.t.sol#L1963-L2022).

smartcontracts commented on packages/contracts-bedrock/test/L1/OPContractsManagerStandardValidator.t.sol:
```diff
@@ -102,6 +103,10 @@ contract OPContractsManagerStandardValidator_TestInit is CommonTest {
 
+        // Skip V1 StandardValidator tests when V2 dispute games are enabled
+        // TODO(#17267): Remove skip when V2 dispute game support added to the StandardValidator is implemented
+        skipIfDevFeatureEnabled(DevFeatures.DEPLOY_V2_DISPUTE_GAMES);
```
Seems ok for now, but blocks shipping to prod

mbaxter replied:
Yep - just trying to split up the work.  We have [another ticket](https://github.com/ethereum-optimism/optimism/issues/17267) to resolve this.

smartcontracts commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -430,8 +431,9 @@ contract OPContractsManager_ChainIdToBatchInboxAddress_Test is Test {
         vm.etch(address(superchainConfigProxy), hex"01");
         vm.etch(address(protocolVersionsProxy), hex"01");
 
+        resolveFeaturesFromEnv();
```
Why was this diff required?

mbaxter replied:
It's not required.  But we're running CI with various feature flags enabled, so wanted to make sure the various test contracts respect the feature flags.

mbaxter committed:
Add function documentation for the new opcm param

mbaxter committed:
Cleanup: add test assertion messages

mbaxter committed:
Remove opcm param from runSingle script

mbaxter committed:
Tweak method documentation

mbaxter committed:
Cleanup dead code

mbaxter committed:
Merge branch 'develop' into snevin/issue-17257/creator-pattern-opcm-deploy
